server/matchmaker/algorithm.py

```diff
@@ -87,42 +93,112 @@ def _propose(self, search: Search, preferred: Search):
             self._unmatch(preferred)
             self._match(search, preferred)
 
-    def _match(self, s1: Search, s2: Search):
-        self._logger.debug("Matching %s and %s", s1, s2)
-        self.matches[s1] = s2
-        self.matches[s2] = s1
 
-    def _unmatch(self, s1: Search):
-        s2 = self.matches[s1]
-        self._logger.debug("Unmatching %s and %s", s1, s2)
-        assert self.matches[s2] == s1
-        del self.matches[s1]
-        del self.matches[s2]
+class RandomlyMatchNewbies(MatchmakingPolicy):
+    def find(self) -> Dict[Search, Search]:
+        self.matches.clear()
+
+        unmatched_newbies = [
+            search for search in self.searches 
+            if search.is_single_ladder_newbie()
+            and not search in self.matches
+        ] 
+
+        while len(unmatched_newbies) >= 2:
+            newbie1 = unmatched_newbies.pop()
+            newbie2 = unmatched_newbies.pop()
+            self._match(newbie1, newbie2)
+
+        if len(unmatched_newbies) == 1:
+            newbie = unmatched_newbies[0]
+
+            default_if_no_available_opponent = None
 
+            opponent = next(
+                (search for search in self.searches
+                if search != newbie
+                and not search in self.matches
+                and search.is_single_party()
+                and search.has_no_top_player()
+                ),
+                default_if_no_available_opponent
+            )
+            if opponent is not default_if_no_available_opponent:
+                self._match(newbie, opponent)
+
+        return self.matches
+
+
+class Matchmaker(object):
+    def __init__(self, searches: List[Search]):
+        self.searches = searches
+        self.matches: Dict[Search, Search] = {}
+
+    def find(self) -> List[Match]:
+        self.matches.update(StableMarriage(self.searches).find())
 
-def _rank_all(searches: List[Search]) -> Dict[Search, List[Search]]:
-    """ Returns searches with best quality for each search.
+        remaining_searches = [
+                search for search in self.searches 
```

Comment:
I think there's an extra indent here
```suggestion
            search for search in self.searches 
```